src/Symfony/Bridge/Monolog/Handler/SwiftMailerHandler.php

```diff
@@ -12,82 +12,162 @@
+        public function onCliTerminate(ConsoleTerminateEvent $event)
+        {
+            $this->instantFlush = true;
+        }
+
+        /**
+         * {@inheritdoc}
+         */
+        protected function send($content, array $records)
+        {
+            parent::send($content, $records);
+
+            if ($this->instantFlush) {
+                $this->flushMemorySpool();
+            }
+        }
+
+        /**
+         * {@inheritdoc}
+         */
+        public function reset()
+        {
             $this->flushMemorySpool();
         }
-    }
 
-    /**
-     * {@inheritdoc}
-     */
-    public function reset()
-    {
-        $this->flushMemorySpool();
-    }
+        /**
+         * Flushes the mail queue if a memory spool is used.
+         */
+        private function flushMemorySpool()
+        {
+            $mailerTransport = $this->mailer->getTransport();
+            if (!$mailerTransport instanceof \Swift_Transport_SpoolTransport) {
+                return;
+            }
 
+            $spool = $mailerTransport->getSpool();
+            if (!$spool instanceof \Swift_MemorySpool) {
+                return;
+            }
+
+            if (null === $this->transport) {
+                throw new \Exception('No transport available to flush mail queue');
+            }
+
+            $spool->flushQueue($this->transport);
+        }
+    }
+} else {
     /**
-     * Flushes the mail queue if a memory spool is used.
+     * Extended SwiftMailerHandler that flushes mail queue if necessary.
+     *
+     * @author Philipp Kräutli <pkraeutli@astina.ch>
+     *
+     * @final  since Symfony 4.3
      */
-    private function flushMemorySpool()
+    class SwiftMailerHandler extends BaseSwiftMailerHandler
     {
-        $mailerTransport = $this->mailer->getTransport();
-        if (!$mailerTransport instanceof \Swift_Transport_SpoolTransport) {
-            return;
+        protected $transport;
+
+        protected $instantFlush = false;
+
+        public function setTransport(\Swift_Transport $transport)
+        {
+            $this->transport = $transport;
+        }
+
+        /**
+         * After the kernel has been terminated we will always flush messages.
+         */
+        public function onKernelTerminate(PostResponseEvent $event)
+        {
+            $this->instantFlush = true;
         }
 
-        $spool = $mailerTransport->getSpool();
-        if (!$spool instanceof \Swift_MemorySpool) {
-            return;
+        /**
+         * After the CLI application has been terminated we will always flush messages.
+         */
+        public function onCliTerminate(ConsoleTerminateEvent $event)
+        {
+            $this->instantFlush = true;
         }
 
-        if (null === $this->transport) {
-            throw new \Exception('No transport available to flush mail queue');
+        /**
+         * {@inheritdoc}
+         */
+        protected function send(string $content, array $records): void
```

Comment:
Maybe we can create an abstract class where everything but this change resides, extend that and then only duplicate this function instead. This will keep the change smaller and hopefully make it clearer why why the duplication exists

Reply:
Yes, I could do that. But then I would need to make `flushMemorySpool` protected. Not sure if that's an improvement. 😕